libdrgn/python/helpers.c

```diff
@@ -249,3 +250,267 @@ PyObject *drgnpy_linux_helper_pgtable_l5_enabled(PyObject *self, PyObject *args,
+	struct prog_or_ns_arg prog_or_ns;
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&:for_each_task",
+					 keywords, &prog_or_pid_ns_converter,
+					 &prog_or_ns))
+		return NULL;
+
+	GenericIterator *iterator =
+		(GenericIterator *)GenericIterator_type.tp_alloc(
+			&GenericIterator_type, 0);
+	if (!iterator)
+		goto out;
+	iterator->prog = NULL;
+	iterator->next = for_each_task_next;
+	iterator->iterator_deinit = (void (*)(void *))task_iter_deinit;
+	iterator->iterator = malloc(sizeof(struct task_iter));
+	if (!iterator->iterator) {
+		err = &drgn_enomem;
+		goto out;
+	}
+	err = task_iter_init(iterator->iterator, prog_or_ns.ns);
+out:
+	prog_or_ns_cleanup(&prog_or_ns);
+	if (err || !iterator) {
+		Py_XDECREF(iterator);
+		return err ? set_drgn_error(err) : NULL;
+	}
+	iterator->prog = prog_or_ns.prog;
+	Py_INCREF(iterator->prog);
+	return iterator;
+}
+
+static PyObject *for_each_pid_next(GenericIterator *self)
+{
+	struct drgn_error *err;
+	struct pid_iter *iterator = (struct pid_iter *)self->iterator;
+	struct drgn_object *entry;
+	err = pid_iter_next(iterator, &entry);
+	if (err)
+		return set_drgn_error(err);
+	if (!entry)
+		return NULL;
+	DrgnObject *ret = DrgnObject_alloc(self->prog);
+	if (!ret)
+		return NULL;
+	err = drgn_object_copy(&ret->obj, entry);
+	if (err)
+		return set_drgn_error(err);
+	return (PyObject *)ret;
+}
+
+GenericIterator *
+drgnpy_linux_helper_for_each_pid(PyObject *self, PyObject *args, PyObject *kwds)
+{
+	static char *keywords[] = {"prog_or_ns", NULL};
+	struct drgn_error *err = NULL;
+	struct prog_or_ns_arg prog_or_ns;
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&:for_each_pid",
+					 keywords, &prog_or_pid_ns_converter,
+					 &prog_or_ns))
+		return NULL;
+
+	GenericIterator *iterator =
+		(GenericIterator *)GenericIterator_type.tp_alloc(
+			&GenericIterator_type, 0);
+	if (!iterator)
+		goto out;
+	iterator->prog = NULL;
+	iterator->next = for_each_pid_next;
+	iterator->iterator_deinit = (void (*)(void *))pid_iter_deinit;
+	iterator->iterator = malloc(sizeof(struct pid_iter));
+	if (!iterator->iterator) {
+		err = &drgn_enomem;
+		goto out;
+	}
+	err = pid_iter_init(iterator->iterator, prog_or_ns.ns);
+out:
+	prog_or_ns_cleanup(&prog_or_ns);
+	if (err || !iterator) {
+		Py_XDECREF(iterator);
+		return set_drgn_error(err);
+	}
+	iterator->prog = prog_or_ns.prog;
+	Py_INCREF(iterator->prog);
+	return iterator;
+}
+
+static PyObject *idr_iter_entry_wrap(struct radix_tree_iter_entry *entry, Program *prog)
+{
+	PyObject *index = PyLong_FromUnsignedLong(entry->index);
+	if (!index)
+		return NULL;
+	DrgnObject *node = DrgnObject_alloc(prog);
+	if (!node) {
+		Py_DECREF(index);
+		return NULL;
+	}
+	struct drgn_error* err = drgn_object_copy(&node->obj, &entry->node);
+	if (err) {
+		Py_DECREF(index);
+		Py_DECREF(node);
+		return set_drgn_error(err);
+	}
+	PyObject *ret = PyTuple_Pack(2, index, node);
+	if (!ret) {
+		Py_DECREF(index);
+		Py_DECREF(node);
+		return NULL;
+	}
+	return ret;
```

Comment:
A few things here:

* `unsigned long` is not large enough for a `uint64_t` on 32-bit systems.
* You can use `Py_BuildValue` instead of manually creating a `PyLong`.
* The tuple grabs a reference on its elements, so you need to drop your original reference.
Untested, but I think this would work:

```suggestion
	DrgnObject *node = DrgnObject_alloc(prog);
	if (!node) {
		Py_DECREF(index);
		return NULL;
	}
	struct drgn_error *err = drgn_object_copy(&node->obj, &entry->node);
	if (err) {
		Py_DECREF(node);
		return set_drgn_error(err);
	}
	PyObject *ret = Py_BuildValue("KO", (unsigned long long)entry->index, node);
	Py_DECREF(node);
	return ret;
```